meson: Fix cairo_ft_dep when fontconfig is a subproject

xclaesse opened the pull request:
When cairo comes from a subproject fallback it should still take
    cairo-ft dependency, otherwise it miss freetype2 cflags.
    
    Also update wrap file to use upstream repository now that meson support
    has been merged.
    
    This requires Meson >= 0.54.0 because cairo itself requires that version
    and we use the new `fallback: 'cairo'` syntax without specifying a
    variable name because cairo does not provide a variable name for
    cairo-ft.

khaledhosny commented:
I’m feeling really uneasy about the growing complexity of the fallback dependencies. HarfBuzz build system should not be a mini package manager. I’m inclined to completely drop it and let the users handle the dependencies as they see fit.

alerque commented:
> I’m feeling really uneasy about the growing complexity of the fallback dependencies. HarfBuzz build system should not be a mini package manager. I’m inclined to completely drop it and let the users handle the dependencies as they see fit.

Honestly I don't really make the connection between declaring dependencies (however complex they are) using a build system and becoming a package manager.

...that being said I do agree that the series of fallbacks is adding a lot of complexity we might not need. I think it also adds quite a bit of unpredictability. I would tend to favor explicit build flags, and the user setting those flags would either enable a feature and *enforce* it's dependencies or disable it and not look for them at all. With so many automatic detection and fallbacks, two people can build the same thing on relatively similar platforms and not come away with the same library at all. If there was a default baseline config and builders had to explicitly set anything they wanted differently for their platform and what dependencies they wanted to pull in or not things would be a bit more predictable.

But this PR is fine ... this is discussion for future direction. I wouldn't suggest redoing anything in the build stack to hastily, and certainly fixing issues like this is fine for now.

khaledhosny commented:
> Honestly I don't really make the connection between declaring dependencies (however complex they are) using a build system and becoming a package manager.

This is done for ”fallback” dependency, which in meson speak a subproject that we build when the dependency is not found. Now we depend on `cairo-ft`, it is not our job to also mange `cairo-ft` own dependencies, which is what is being done here.

khaledhosny commented:
There is also no explanation what this change is fixing, only what is doing.

alerque commented:
Okay you might be right.

xclaesse commented:
The problem here is `cairo_ft_dep = cairo_dep` is wrong because it miss cflags for fontconfig. It is true that all harfbuzz should only care about is getting `cairo-ft` dependency, but unfortunately cairo's meson port is missing that, it's not even generating cairo-ft.pc. That's a bug in cairo.

The proper fix should be done in both cairo in harfbuzz:
- Cairo meson should provide cairo-ft dependency
- Harfbuzz should always do `cairo_ft_dep = dependency('cairo-ft', required: get_option('cairo'))` regardless if cairo comes from system or subproject.

Side question: Does harfbuzz requires that cairo has been built with fontconfig? Or can it work either way?

khaledhosny commented:
> The proper fix should be done in both cairo in harfbuzz:

Agree. We will also need to switch to Cairo upstream not the fork we are using now.

>  Does harfbuzz requires that cairo has been built with fontconfig

We don’t seem to use any of Cairo’s FontConfig function, and we check only for cairo-ft.

xclaesse commented:
> We don’t seem to use any of Cairo’s FontConfig function, and we check only for cairo-ft.

Seems I got this wrong, cairo-ft is for FreeType and not FontConfig. Sorry for the confusion.

I created that cairo PR: https://gitlab.freedesktop.org/cairo/cairo/-/merge_requests/68

xclaesse commented:
Sorry for the confusing. I updated this PR to use the fix I did cairo and updated description and commit message. Hopfully it makes things a bit more clear.

xclaesse commented:
CC @tp-m

xclaesse commented:
Hmm this requires Meson 0.54 in the CI.

behdad commented:
I share Khaled's concerns.

tp-m commented:
As far as I can tell this MR is a clean-up of some meson code plus a correctness fix so that harfbuzz can be consumed correctly as subproject by other projects such as cairo, Gtk, GStreamer in certain situations.

This is independent of the .wrap files in `harfbuzz/subprojects`. Those are for convenience, and they make it very easy to build and hack on harfbuzz and its stack in some situations, e.g. on windows or older distros or when doing fuzzing/CI. You can remove these wrap files of course, but I think they're quite cool to have, and shouldn't be too much hassle once things are settled. Removing them will only make it more difficult to build harfbuzz.

The fix would be needed even if you decided not to ship wrap files in harfbuzz any more. The fix is for other projects that depend on you.

The cairo meson port landed upstream only recently, and we're still shaking out some bugs, sorry about that!

xclaesse commented:
Right, sorry for all the confusion here, my initial patch was wrong and the right fix for the problem I had is actually in cairo: https://gitlab.freedesktop.org/cairo/cairo/-/merge_requests/68/diffs?commit_id=4395f7a5b60004bfe1ac62c2bae993385cb8b504.

Since then I updated this PR (probably should have closed and opened a new one) with a cleanup how harfbuzz lookup the cairo dependency. It's actually simplifying the code (+12/-23 lines).

alerque commented:
No worries on the PR update, GitHub's tooling actually makes it easier to review this way than closing/re-opening PRs. There is no easy way to compare what changed between PRs, but there is a way to see what changed since your last visit to a PR and also see where the force-pushed commits fit into the discussion timeline.

These changes are definitely looking better! My only question is will this be a regression for building against systems that don't have the latest Cairo with the upstream build fixes?

xclaesse commented:
> These changes are definitely looking better! My only question is will this be a regression for building against systems that don't have the latest Cairo with the upstream build fixes?

This PR should make no difference when getting cairo from system. 

> Right, sorry for all the confusion here, my initial patch was wrong and the right fix for the problem I had is actually in cairo: https://gitlab.freedesktop.org/cairo/cairo/-/merge_requests/68/diffs?commit_id=4395f7a5b60004bfe1ac62c2bae993385cb8b504.

And now I realize even that patch is wrong, cairo.pc is supposed to require fontconfig already, so it's not needed in cairo-ft.pc. But it's not harfbuzz issue anyway...

xclaesse commented:
@behdad @tp-m @alerque The cairo side of this has been merged, any objection to merge this in harfbuzz? I updated the PR to not require a newer version of Meson.

xclaesse commented:
Argh, cairo requires meson >= 0.54.0, so it needs to be updated in the CI too since the CI uses cairo as subproject.

alerque commented:
> Argh, cairo requires meson >= 0.54.0

Meson distribution is still very uneven. Releases with different bugs have meant some platforms have gotten stuck upgrading on various different releases. Given earlier troubles and the effort that went into getting the current build system working across older meson releases (and the unresolved tension with deprecating other build systems) I don't think this is a good juncture to narrow the range of supported meson versions.

xclaesse commented:
I'm not aware of bugs in meson affecting harfbuzz, can you please be more specific and report issues of any? It is true that. 0 release have their regressions but there are point releases that fix most reported issues.

xclaesse commented:
Who knows how to update meson version in CI runners?

khaledhosny commented:
We are intentionally requiring meson >= 0.47.0, we do not wish to require a newer version of meson for the time being. Some of the CI jobs intentionally install meson==0.47.0 to enforce this.

xclaesse commented:
@khaledhosny why requiring a very old version? Meson support just got merged in harfbuzz very recently, users on very old system won't use it any way.  In the early days it's better to require latest version to not accumulate legacy.

khaledhosny commented:
https://github.com/harfbuzz/harfbuzz/pull/2632#issuecomment-672265606

xclaesse commented:
I strongly disagree, but if you prefer broken Cairo detection that's your call. I guess you'll be able to merge this in about 2 or 3 years.

alerque commented:
> Meson support just got merged in harfbuzz very recently, users on very old system won't use it any way.

The hope with Meson support is that we would be able to *drop* the other build systems. Currently there are 3: automake, cmake, and meson. You are correct that Meson was just recently added, but the others were (a bit hastily) removed entirely. Then they were fetched back because it became clear that Meson didn't work for everybody and we needed to support a longer transition period. At that point a huge effort was made to bring the Meson support from only working on very new builds to working on much older Meson. If and when we do keep moving in the direction of dropping cmake and/or automake, having support for a broad range of Meson versions will be crucial. Harfbuzz deploys in a *lot* of places with a wide variety of system tooling. Requiring very new build tooling would be a huge regression and work against keeping the Harfbuzz ecosystem as up to date as possible. As long as people can keep dropping new Harfbuzz versions into old systems without breakage those old systems stand a chance of getting good HB support. As soon as you require fresh dependencies or tooling you start limiting what platforms can keep up.

> I strongly disagree, but if you prefer broken Cairo detection that's your call. I guess you'll be able to merge this in about 2 or 3 years.

I think we'd prefer broken Cairo detection to *requiring* the freshest Meson. But is that really a compromise we have to make? Is there really no other way? I believe there must be work-arounds so we can continue to support building under older Meson so that we can keep moving forward with deprecating other build systems.

xclaesse commented:
Hmmm ok, sorry for having been harsh, I miss understood the linked comment, I thought you wanted to keep autotools indefinitely. If the plan is to be meson only then it makes much more sense.

What I could do is do the cairo fallback only if meson version is new enough. With older version of meson it will only look on system.

xclaesse commented:
@alerque problem remains that the Windows CI requires cairo as subproject. Maybe we can update meson version only for that CI job?

QuLogic commented:
FYI, this fixes building the full GStreamer set, with their overarching `gst-build` project, on Windows. On current `master`, it fails to build due to missing includes (presumably from missing search paths.)

khaledhosny commented:
> problem remains that the Windows CI requires cairo as subproject. Maybe we can update meson version only for that CI job?

This should be fine since this CI job is also testing the use of subprojects.

xclaesse commented:
Hmm build of cairo fails:
```
FAILED: subprojects/cairo/src/libcairo-2.dll 
i686-w64-mingw32-gcc  -o subprojects/cairo/src/libcairo-2.dll 'subprojects/cairo/src/da09890@@cairo@sha/cairo-analysis-surface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-arc.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-array.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-atomic.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-base64-stream.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-base85-stream.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-bentley-ottmann-rectangular.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-bentley-ottmann-rectilinear.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-bentley-ottmann.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-botor-scan-converter.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-boxes-intersect.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-boxes.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-cache.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-clip-boxes.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-clip-polygon.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-clip-region.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-clip-surface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-clip-tor-scan-converter.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-clip.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-color.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-composite-rectangles.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-compositor.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-contour.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-damage.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-debug.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-default-context.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-device.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-error.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-fallback-compositor.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-fixed.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-font-face-twin-data.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-font-face-twin.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-font-face.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-font-options.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-freed-pool.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-freelist.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-gstate.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-hash.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-hull.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-image-compositor.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-image-info.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-image-source.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-image-surface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-line.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-lzw.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-mask-compositor.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-matrix.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-mempool.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-mesh-pattern-rasterizer.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-misc.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-mono-scan-converter.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-mutex.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-no-compositor.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-observer.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-output-stream.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-paginated-surface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-path-bounds.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-path-fill.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-path-fixed.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-path-in-fill.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-path-stroke-boxes.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-path-stroke-polygon.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-path-stroke-traps.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-path-stroke-tristrip.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-path-stroke.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-path.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-pattern.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-pen.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-polygon-intersect.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-polygon-reduce.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-polygon.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-raster-source-pattern.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-recording-surface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-rectangle.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-rectangular-scan-converter.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-region.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-rtree.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-scaled-font.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-shape-mask-compositor.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-slope.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-spans-compositor.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-spans.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-spline.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-stroke-dash.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-stroke-style.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-surface-clipper.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-surface-fallback.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-surface-observer.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-surface-offset.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-surface-snapshot.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-surface-subsurface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-surface-wrapper.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-surface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-time.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-tor-scan-converter.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-tor22-scan-converter.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-toy-font-face.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-traps-compositor.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-traps.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-tristrip.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-unicode.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-user-font.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-version.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-wideint.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-cff-subset.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-scaled-font-subsets.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-truetype-subset.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-type1-fallback.c.obj' 'subprojects/cairo/src/da09890@
@cairo@sha/cairo-type1-glyph-names.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-type1-subset.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-type3-glyph-surface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-pdf-operators.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-pdf-shading.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-tag-attributes.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-deflate-stream.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-png.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-svg-surface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-ft-font.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/win32_cairo-win32-debug.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/win32_cairo-win32-device.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/win32_cairo-win32-gdi-compositor.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/win32_cairo-win32-system.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/win32_cairo-win32-surface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/win32_cairo-win32-display-surface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/win32_cairo-win32-printing-surface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/win32_cairo-win32-font.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-composite.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-device.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-dispatch.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-glyphs.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-gradient.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-info.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-msaa-compositor.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-operand.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-shaders.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-source.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-spans-compositor.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-surface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-traps-compositor.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-wgl-context.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-script-surface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-ps-surface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-pdf-surface.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-pdf-interchange.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-tag-stack.c.obj' 'subprojects/cairo/src/da09890@@cairo@sha/cairo-xml-surface.c.obj' -Wl,--allow-shlib-undefined -Wl,-O1 -shared -Wl,--start-group -Wl,--out-implib=subprojects/cairo/src/libcairo.dll.a -static-libgcc -Wl,-Bstatic -lpthread subprojects/zlib/libz.dll.a subprojects/libpng-1.6.35/libpng16.dll.a subprojects/freetype2/libfreetype.dll.a subprojects/pixman/pixman/libpixman-1.dll.a -lm -lgdi32 -lmsimg32 -lopengl32 -lkernel32 -luser32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -Wl,--end-group
/usr/bin/i686-w64-mingw32-ld: subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-composite.c.obj:cairo-gl-composite.c:(.text+0x156): undefined reference to `glActiveTexture@4'
/usr/bin/i686-w64-mingw32-ld: subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-composite.c.obj:cairo-gl-composite.c:(.text+0x252): undefined reference to `glActiveTexture@4'
/usr/bin/i686-w64-mingw32-ld: subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-composite.c.obj:cairo-gl-composite.c:(.text+0xb0a): undefined reference to `glBlendFuncSeparate@16'
/usr/bin/i686-w64-mingw32-ld: subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-composite.c.obj:cairo-gl-composite.c:(.text+0xbd0): undefined reference to `glBlendFuncSeparate@16'
/usr/bin/i686-w64-mingw32-ld: subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-device.c.obj:cairo-gl-device.c:(.text+0xaaf): undefined reference to `glActiveTexture@4'
/usr/bin/i686-w64-mingw32-ld: subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-device.c.obj:cairo-gl-device.c:(.text+0xaca): undefined reference to `glActiveTexture@4'
/usr/bin/i686-w64-mingw32-ld: subprojects/cairo/src/da09890@@cairo@sha/cairo-gl-glyphs.c.obj:cairo-gl-glyphs.c:(.text+0xc1): undefined reference to `glActiveTexture@4'
collect2: error: ld returned 1 exit status
```

CC @tp-m @nirbheek

xclaesse commented:
@alatiera Added a commit here to update all wraps while at it. This should replace https://github.com/harfbuzz/harfbuzz/pull/2746.

Remains that cairo cannot be cross built to windows currently, I reported the bug there if someone wants to investigate: https://gitlab.freedesktop.org/cairo/cairo/-/issues/435.

xclaesse commented:
Ok, it seems the cairo issue has been already fixed, or at least worked around enough that the CI now pass here. So I think this is ready to be merged now.

khaledhosny commented:
Thanks! I think freetype needs to be updated as well since upstream has meson support now.

xclaesse commented:
> Thanks! I think freetype needs to be updated as well since upstream has meson support now.

Indeed, but the meson port that got merged upstream got developed independently to the port we are using here, and it has a few issues. They are in the process of migrating to gitlab which will make it much easier to contribute, I made a MR already: https://gitlab.freedesktop.org/freetype/freetype/-/merge_requests/1
